packages/expo-updates/ios/EXUpdates/EXUpdatesErrorRecovery.m

```diff
@@ -0,0 +1,220 @@
+//  Copyright © 2021 650 Industries. All rights reserved.
+
+#import <EXUpdates/EXUpdatesAppLauncherWithDatabase.h>
+#import <EXUpdates/EXUpdatesErrorRecovery.h>
+#import <React/RCTAssert.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+typedef NS_ENUM(NSInteger, EXUpdatesErrorRecoveryTask) {
+  EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate,
+  EXUpdatesErrorRecoveryTaskLaunchNew,
+  EXUpdatesErrorRecoveryTaskLaunchCached,
+  EXUpdatesErrorRecoveryTaskLaunchEmbedded,
+  EXUpdatesErrorRecoveryTaskCrash
+};
+
+static NSInteger const EXUpdatesErrorRecoveryRemoteLoadTimeoutMs = 5000;
+
+@interface EXUpdatesErrorRecovery ()
+
+@property (nonatomic, strong) NSMutableArray<NSNumber *> *pipeline;
+@property (nonatomic, assign) BOOL isRunning;
+@property (nonatomic, assign) BOOL isWaitingForRemoteUpdate;
+@property (nonatomic, assign) NSInteger remoteLoadTimeout;
+
+@property (nonatomic, strong) dispatch_queue_t errorRecoveryQueue;
+
+@property (nonatomic, strong) NSMutableArray *encounteredErrors;
+
+@end
+
+@implementation EXUpdatesErrorRecovery
+
+- (instancetype)init
+{
+  return [self initWithErrorRecoveryQueue:dispatch_queue_create("expo.controller.errorRecoveryQueue", DISPATCH_QUEUE_SERIAL)
+                        remoteLoadTimeout:EXUpdatesErrorRecoveryRemoteLoadTimeoutMs];
+}
+
+- (instancetype)initWithErrorRecoveryQueue:(dispatch_queue_t)errorRecoveryQueue
+                         remoteLoadTimeout:(NSInteger)remoteLoadTimeout
+{
+  if (self = [super init]) {
+    _pipeline = @[
+      @(EXUpdatesErrorRecoveryTaskWaitForRemoteUpdate),
+      @(EXUpdatesErrorRecoveryTaskLaunchCached),
+      @(EXUpdatesErrorRecoveryTaskLaunchEmbedded),
+      @(EXUpdatesErrorRecoveryTaskCrash)
+    ].mutableCopy;
+    _isRunning = NO;
+    _isWaitingForRemoteUpdate = NO;
+    _errorRecoveryQueue = errorRecoveryQueue;
+    _remoteLoadTimeout = remoteLoadTimeout;
+    _encounteredErrors = [NSMutableArray new];
+  }
+  return self;
+}
+
+- (void)handleError:(NSError *)error fromLaunchedUpdate:(nullable EXUpdatesUpdate *)launchedUpdate
+{
+  if (launchedUpdate) {
+    [self _markUpdateFailed:launchedUpdate];
+  }
+  [self _startPipelineWithEncounteredError:error];
```

Comment:
Since _markUpdateFailed dispatches to the DB queue and then potentially dispatches to the error recovery queue, it's likely that _startPipelineWithEncounteredError will dispatch to the error recovery queue first. This creates a scenario where _markUpdateFailed removes EXUpdatesErrorRecoveryTaskLaunchCached from the task queue too late.

Possible fix is to give _markUpdateFailed a completion block.

Reply:
Based on our conversation last week I decided this isn't worth the added complexity. If this database call fails, either the entire database is corrupt (in which case we'll emergency launch the embedded update next time), or this one call fails, in which case the worst thing that happens is we'll try to launch it again next time. The main priority with this whole feature is to allow the app to fix itself by downloading _new_ updates whenever possible, so I don't think this edge case with rolling back is worth about here. (I could be convinced otherwise)

(also this function has moved to https://github.com/expo/expo/pull/14398 -- https://github.com/expo/expo/pull/14398/files#diff-34b4cfcbaf8386b6601d2b279f4c38b67bbd8c5e962e276677a6176b879ac024R366)